pkg/rootless/rootless.go

```diff
@@ -0,0 +1,126 @@
+package rootless
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"runtime"
+
+	"github.com/containers/storage/pkg/idtools"
+	"github.com/pkg/errors"
+)
+
+/*
+extern int reexec_in_user_namespace(int ready);
+extern int reexec_in_user_namespace_wait(int pid);
+*/
+import "C"
+
+func runInUser() error {
+	os.Setenv("_LIBPOD_USERNS_CONFIGURED", "done")
+	return nil
+}
+
+// IsRootless tells us if we are running in rootless mode
+func IsRootless() bool {
+	return os.Getuid() != 0 || os.Getenv("_LIBPOD_USERNS_CONFIGURED") != ""
+}
+
+func tryMappingTool(tool string, pid int, hostID int, mappings []idtools.IDMap) error {
+	path, err := exec.LookPath(tool)
+	if err != nil {
+		return err
+	}
+
+	appendTriplet := func(l []string, a, b, c int) []string {
+		return append(l, fmt.Sprintf("%d", a), fmt.Sprintf("%d", b), fmt.Sprintf("%d", c))
+	}
+
+	args := []string{path, fmt.Sprintf("%d", pid)}
+	args = appendTriplet(args, 0, hostID, 1)
+	if mappings != nil {
+		for _, i := range mappings {
+			args = appendTriplet(args, i.ContainerID+1, i.HostID, i.Size)
+		}
+	}
+	cmd := exec.Cmd{
+		Path: path,
+		Args: args,
+	}
+	return cmd.Run()
+}
+
+// BecomeRootInUserNS re-exec podman in a new userNS
+func BecomeRootInUserNS() (bool, error) {
+	runtime.LockOSThread()
+
+	if os.Getuid() == 0 || os.Getenv("_LIBPOD_USERNS_CONFIGURED") != "" {
```

Comment:
Should this happen before `LockOSThread()`? Seems like it doesn't need to be guaranteed to run in a single thread, and this should be the typical case, so we can save some time by not calling `LockOSThread()` in this case

Reply:
changed